File: src/main/java/io/tatum/model/request/CreateTransaction.java
package io.tatum.model.request;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import javax.validation.constraints.*;
import java.math.BigDecimal;

/**
 * The type Create transaction.
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@EqualsAndHashCode
public class CreateTransaction {

    @NotEmpty
    @Size(min = 24, max = 24)
    private String senderAccountId;

    @NotEmpty
    @Size(min = 24, max = 24)
    private String recipientAccountId;

    @NotEmpty
    @Size(max = 38)
    @Pattern(regexp = "^[+]?((\\d+(\\.\\d*)?)|(\\.\\d+))$") // number string
    private String amount;

    @Size(min = 1, max = 100)
    private String paymentId;

    @Size(min = 1, max = 100)
    private String transactionCode;

    @Size(min = 1, max = 500)
    private String senderNote;

    @Size(min = 1, max = 500)
    private String recipientNote;

    @Min(0)
    private BigDecimal baseRate;

    private Boolean anonymous;

    private Boolean compliant;

}

File: src/main/kotlin/io/tatum/model/request/CreateTransaction.kt
package io.tatum.model.request

import java.math.BigDecimal

class CreateTransaction(
    val senderAccountId: String,
    val recipientAccountId: String,
    amount: BigDecimal,
    val paymentId: String? = null,
    val transactionCode: String? = null,
    val senderNote: String? = null,
    val recipientNote: String? = null,
    val baseRate: BigDecimal? = null,
    val anonymous: Boolean?,
    val compliant: Boolean?,
) {
    init {
        require(senderAccountId.length == 24)
        require(recipientAccountId.length == 24)
        require(amount.toString().length <= 38)
        if (paymentId != null) require(paymentId.length in 1..100)
        if (transactionCode != null) require(transactionCode.length in 1..100)
        if (senderNote != null) require(senderNote.length in 1..100)
        if (recipientNote != null) require(recipientNote.length in 1..100)
        if (baseRate != null) require(baseRate >= BigDecimal.ZERO)
    }
    val amount: String = amount.toString()
}
